python/lsst/meas/extensions/shapeHSM/_hsm_higher_moments.py

```diff
@@ -0,0 +1,382 @@
+
+    def _calculate_higher_order_moments(
+        self,
+        image,
+        center,
+        M,
+        badpix=None,
+        set_masked_pixels_to_zero=False,
+        use_linear_algebra=False,
+    ):
+        """
+        Calculate the higher order moments of an image.
+
+        Parameters
+        ----------
+        image : `~lsst.afw.image.Image`
+            Image of the PSF
+        center: `~lsst.geom.Point2D`
+            First order moments of ``image``. This is used as the peak of the
+            Gaussian weight image.
+        M : `~numpy.ndarray`
+            A 2x2 numpy array representing the second order moments of
+            ``image``. This is used to generate the Gaussian weight image.
+        badpix : `~numpy.ndarray` or None
+            A 2D array having the same shape and orientation as ``image.array``
+            that denotes which pixels are bad and should not be accounted for
+            when computing the moments.
+        set_masked_pixels_to_zero: `bool`
+            Whether to treat pixels corresponding to ``badpix`` should be set
+            to zero, or replaced by a scaled version of the weight image.
+            This is ignored if ``badpix`` is None.
+        use_linear_algebra: `bool`
+            Use linear algebra operations (eigen decomposition and inverse) to
+            calculate the moments? If False, use the specialized formulae for
+            2x2 matrix.
+
+        Returns
+        -------
+        results : `dict`
+            A dictionary mapping the order of the moments expressed as tuples
+            to the corresponding higher order moments.
+        """
+
+        bbox = image.getBBox()
+        image_array = image.array
+
+        y, x = np.mgrid[: image_array.shape[0], : image_array.shape[1]]
+
+        if use_linear_algebra:
+            inv_M = np.linalg.inv(M)
+
+            evalues, evectors = np.linalg.eig(inv_M)
+
+            sqrt_inv_M = evectors * np.sqrt(evalues) @ np.linalg.inv(evectors)
+        else:
+            D = M[0, 0] * M[1, 1] - M[0, 1] * M[1, 0]
+            sqrt_D = D**0.5
+            sqrt_eta = (D * (M[0, 0] + M[1, 1] + 2 * sqrt_D)) ** 0.5
+            sqrt_inv_M = (1 / sqrt_eta) * np.array(
+                [[M[1, 1] + sqrt_D, -M[0, 1]], [-M[1, 0], M[0, 0] + sqrt_D]]
+            )
+
+        pos = np.array([x - (center.getX() - bbox.getMinX()), y - (center.getY() - bbox.getMinY())])
+
+        std_pos = np.einsum("ij,jqp->iqp", sqrt_inv_M, pos)
+        weight = np.exp(-0.5 * np.einsum("ijk,ijk->jk", std_pos, std_pos))
+
+        std_x, std_y = std_pos
+
+        if badpix is not None and badpix.any():
+            if set_masked_pixels_to_zero:
+                image_array[badpix] = 0.0
+            else:
+                image_array[badpix] = weight[badpix] * image_array[~badpix].sum() / weight[~badpix].sum()
+
+        image_weight = weight * image_array
+        normalization = np.sum(image_weight)
+
+        results = {}
```

Comment:
Just saw the `use_linear_algebra` bit appear, so seems you're staring at this too :).

Aside from the 2x2 specialization (which I think should be the only code path BTW), I got a tad bit more performance by caching the power arrays.  Something like:

```python
    std_x_pow = {}
    std_y_pow = {}

    pmin = min(p for p, q in pqlist)
    qmin = min(q for p, q in pqlist)
    if pmin == 0:
        std_x_pow[0] = 1.0
    elif pmin == 1:
        std_x_pow[1] = std_x
    else:
        std_x_pow[pmin] = std_x**pmin

    if qmin == 0:
        std_y_pow[0] = 1.0
    elif qmin == 1:
        std_y_pow[1] = std_y
    else:
        std_y_pow[qmin] = std_y**qmin

    for p in range(pmin, max(p for p, q in pqlist)+1):
        std_x_pow[p+1] = std_x_pow[p] * std_x
    for q in range(qmin, max(q for p, q in pqlist)+1):
        std_y_pow[q+1] = std_y_pow[q] * std_y

    results = {}
    for p, q in pqlist:
        tmp = std_x_pow[p] * std_y_pow[q]
        tmp *= image_weight
        results[(p, q)] = np.sum(tmp) / normalization
```

Reply:
Good idea, and I hope this does well even on objects with large footprints.

`pmin` and `qmin` are always going to be 0 given how we generate those tuples. So I am going to remove the `if`-`elif`-`else` conditions.

Reply:
Actually, it occurs to me that `np.vander` may be your friend here (but I didn't try it).

Reply:
I think I'd like to have the option of having the more expensive code path, given that that was what TQ originally used in the previous HSC data, so I consider that validated.  I'd like to have an option to resort to the validated code if something doesn't look right. I could add another unit test to check that both code paths result in consistent answers.